spt/features/vag_searcher.cpp

```diff
@@ -0,0 +1,242 @@
+#include "stdafx.h"
+#if defined(SSDK2007) || defined(SSDK2013)
+#include "vag_searcher.hpp"
+
+#include "interfaces.hpp"
+#include "property_getter.hpp"
+#include "ent_utils.hpp"
+#include "..\utils\game_detection.hpp"
+#include "playerio.hpp"
+#include "signals.hpp"
+#include "..\sptlib-wrapper.hpp"
+#include "..\cvars.hpp"
+
+VagSearcher spt_vag_searcher;
+
+ConVar y_spt_vag_search_portal(
+    "y_spt_vag_search_portal",
+    "overlay",
+    FCVAR_CHEAT,
+    "Chooses the portal for the VAG search. Valid options are overlay/blue/orange/portal index. This is the portal you enter.\n");
+
+CON_COMMAND(y_spt_vag_search, "Search VAG")
+{
+	spt_vag_searcher.StartSearch();
+}
+
+void VagSearcher::StartSearch()
+{
+	if (IsIterating())
+		return;
+
+	if (strcmp(y_spt_vag_search_portal.GetString(), "overlay") == 0)
+	{
+		enter_portal = getPortal(_y_spt_overlay_portal.GetString(), false);
+	}
+	else
+	{
+		enter_portal = getPortal(y_spt_vag_search_portal.GetString(), false);
+	}
+
+	if (!enter_portal)
+	{
+		Msg("Entry portal not found, maybe try using index.\n");
+		return;
+	}
+
+	exit_portal = GetLinkedPortal(enter_portal);
+	if (!exit_portal)
+	{
+		Msg("Exit portal not found, maybe try using index.\n");
+		return;
+	}
+
+	entry_index = enter_portal->entindex();
+	exit_index = exit_portal->entindex();
+
+	if (utils::GetProperty<int>(entry_index - 1, "m_bActivated") == 0
+	    || utils::GetProperty<int>(exit_index - 1, "m_bActivated") == 0)
+	{
+		Msg("Portal not activated.\n");
+		return;
+	}
+
+	StartIterations();
+}
+
+bool VagSearcher::ShouldLoadFeature()
+{
+	return utils::DoesGameLookLikePortal();
+}
+
+void VagSearcher::InitHooks() {}
+
+void VagSearcher::LoadFeature()
+{
+	InitCommand(y_spt_vag_search);
+	InitConcommandBase(y_spt_vag_search_portal);
+	if (TickSignal.Works)
+	{
+		TickSignal.Connect(this, &VagSearcher::OnTick);
+	}
+	if (VagCrashSignal.Works)
+	{
+		VagCrashSignal.Connect(this, &VagSearcher::VagCrashTriggered);
+	}
+}
+
+void VagSearcher::UnloadFeature() {}
+
+void VagSearcher::VagCrashTriggered()
+{
+	crash = true;
+}
+
+void VagSearcher::StartIterations()
+{
+	if (!y_spt_prevent_vag_crash.GetBool())
+	{
+		Msg("Auto enable y_spt_prevent_vag_crash.\n");
+		y_spt_prevent_vag_crash.SetValue(1);
+	}
+
+	entry_origin = utils::GetPortalPosition(enter_portal);
+	auto angle = utils::GetPortalAngles(enter_portal);
+	// angles_to_vec
+	angle *= (M_PI_F / 180);
+	entry_norm = Vector(std::cos(angle[1]) * std::cos(angle[0]),
```

Comment:
There's an SDK function for calculating the vector out of angles, iirc it's called AngleVectors